### PR TITLE
dynamo: support attribute access on tensor subclasses without sources

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/117596

This was needed for Float8Tensor. Before this PR, dynamo would sometimes handle attribute access on tensor subclasses correctly, but it would choke on tensor subclasses with no source (it would fall back to using a `GetAttrVariable` to represent the attribute access, which is a problem if the attribute is a tensor that we later want to call tensor methods on).

I supported two cases:

(1) the attribute is a tensor, which is part of the `attrs` returned by the subclass's `__tensor_flatten__`. This creates a `TensorVariable`
(2) the attribute is a constant, which is part of the constant metadata returned by `__tensor_flatten__`. As per the contract of tensor_flatten, this should be a `ConstantVariable`. It could be possible that we allow non-constant metadata in the future, but we don't support that today.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119948
* #119067
* #118802
* #118671
* #118670
* #118669
* #119947
* #118803
* #118668
* #118667
* #118191
* #118209
* __->__ #117666
* #117667

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_subclasses.py [hunk 1]
```diff
@@ -1007,6 +1014,17 @@ def forward(self):
...
 
+    def test_wrapper_subclass_dynamo_attribute_access_on_intermediate(self):
+        def f(x_subclass):
+            tmp_subclass = torch.add(x, 1)
+            return torch.mul(tmp_subclass._scale, tmp_subclass._constant)
+
+        # should recompile for different scale size when dynamic=False
+        x = ScaledTensor(torch.randn(2, 4), torch.randn(3), constant=2)
+        out_ref = f(x)
+        out_test = torch.compile(f, backend="aot_eager")(x)
```
Comment (drisspg) [inline test/dynamo/test_subclasses.py:1025]: dumbq, do you have to add full_graph=True to ensure that this isn't graphbreaking?

**Comment by ezyang** [comment]
@voznesenskym can you please look at this

**Comment by wanchaol** [comment]
DTensor had a similar issue hopefully it should solve this! https://github.com/pytorch/pytorch/issues/116712

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -173,6 +175,31 @@ def specialize(value: torch.Tensor):
...
+                    from .builder import wrap_fx_proxy
+
+                    return wrap_fx_proxy(
+                        tx=tx, proxy=proxy, example_value=example_value
+                    )
+                else:
+                    # attributes in the ctx returned by tensor_flatten are assumed to be constants
+                    from . import ConstantVariable
+
+                    return ConstantVariable(example_value)
```
Comment (zou3519) [inline torch/_dynamo/variables/tensor.py:202]: can ctx contain lists? lists cannot be ConstantVariable

Comment (zou3519) [reply torch/_dynamo/variables/tensor.py:202]: This probably needs some assertions that ctx is just what we think it is (ints? strings? etc)

**Review by zou3519 — APPROVED** [review]
As discussed offline, a better long-term thing would be: when Dynamo sees that a function returned a Fakeified Tensor Subclass, then it should also look at the attributes on the FakeTensor, and remember they exist somewhere so that future getattr lookups can find them. This should apply to Tensor Subclasses AND regular Tensors (because people can set attributes on regular Tensors).

But what's in this PR seems reasonable

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (zou3519) [reply torch/_dynamo/variables/tensor.py:202]: Actually, can we just send this through builder.py?

**Pull request closed without merging**